include/oneapi/dpl/internal/dynamic_selection_impl/sycl_scheduler.h

```diff
@@ -0,0 +1,136 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _SYCL_SCHEDULER_IMPL_H
+#define _SYCL_SCHEDULER_IMPL_H
+
+#pragma once
+
+#include <CL/sycl.hpp>
+#include "oneapi/dpl/internal/dynamic_selection.h"
+#include "oneapi/dpl/internal/dynamic_selection_impl/scoring_policy_defs.h"
+#include "oneapi/dpl/internal/dynamic_selection_impl/scheduler_defs.h"
+#include "oneapi/dpl/internal/dynamic_selection_impl/concurrent_queue.h"
+
+#include <atomic>
+#include <mutex>
+#include <stdlib.h>
+
+namespace oneapi {
+namespace dpl {
+namespace experimental {
+
+  struct sycl_scheduler {
+
+    using native_resource_t = sycl::queue;
+    using native_sync_t = sycl::event;
+    using execution_resource_t = oneapi::dpl::experimental::nop_execution_resource_t<native_resource_t>;
+    using universe_container_t = std::vector<execution_resource_t>;
+
+    class async_wait_t {
+    public:
+      virtual void wait_for_all() = 0;
+      virtual native_sync_t get_native() const = 0;
+      virtual ~async_wait_t() {}
+    };
+    using waiter_container_t = Queue<async_wait_t *>;
+
+    template<typename PropertyHandle>
+    class async_wait_impl_t : public async_wait_t {
+      PropertyHandle p_;
+      native_sync_t w_;
+      std::chrono::high_resolution_clock::time_point t0_;
+      std::shared_ptr<std::atomic<bool>> wait_reported_;
+    public:
+
+      async_wait_impl_t(PropertyHandle p, sycl::event e) : p_(p), w_(e), t0_(std::chrono::high_resolution_clock::now()),
+                                                           wait_reported_{std::make_shared<std::atomic<bool>>(false)} { };
+
+      native_sync_t get_native() const override {
+        return w_;
+      }
+
+      void wait_for_all() override {
+        w_.wait();
+      }
+    };
+
+    std::mutex global_rank_mutex_;
+    universe_container_t global_rank_;
+    waiter_container_t waiters_;
+
+    sycl_scheduler() { }
+
+    sycl_scheduler(const sycl_scheduler& v) : global_rank_(v.global_rank_), waiters_(v.waiters_) { }
```

Comment:
In practice this is not a problem, because the copy constructor is not called anywhere because of the common reference semantics. But we should remove the copy constructor.